Mathlib/Testing/SlimCheck/Testable.lean

```diff
@@ -0,0 +1,516 @@
+    let proxy := Shrinkable.shrink (x, y - x)
+    proxy.map (λ ⟨⟨fst, snd⟩, ha⟩ => ⟨⟨fst, fst + snd, sorry⟩, sorry⟩)
+```
+Again, we take advantage of the fact that other types have useful
+`Shrinkable` implementations, in this case `Prod`. Note that the second
+proof is heavily based on `SizeOf` since its used for termination so
+the first step you want to take is almost always to `simp` in order to
+get through both `SizeOf` abstraction by `SlimCheck` and the auto generated
+`SizeOf` instances.
+
+## Main definitions
+  * `Testable` class
+  * `Testable.check`: a way to test a proposition using random examples
+
+## Tags
+
+random testing
+
+## References
+  * https://hackage.haskell.org/package/QuickCheck
+-/
+
+namespace SlimCheck
+
+variable {β : α → Prop}
+
+/-- Result of trying to disprove `p`
+The constructors are:
+  *  `success : (PSum Unit p) → TestResult p`
+     succeed when we find another example satisfying `p`
+     In `success h`, `h` is an optional proof of the proposition.
+     Without the proof, all we know is that we found one example
+     where `p` holds. With a proof, the one test was sufficient to
+     prove that `p` holds and we do not need to keep finding examples.
+   * `gaveUp : ℕ → TestResult p`
+     give up when a well-formed example cannot be generated.
+     `gaveUp n` tells us that `n` invalid examples were tried.
+     Above 100, we give up on the proposition and report that we
+     did not find a way to properly test it.
+   * `failure : ¬ p → (List String) → ℕ → TestResult p`
+     a counter-example to `p`; the strings specify values for the relevant variables.
+     `failure h vs n` also carries a proof that `p` does not hold. This way, we can
+     guarantee that there will be no false positive. The last component, `n`,
+     is the number of times that the counter-example was shrunk.
+-/
+inductive TestResult (p : Prop) where
+| success : PSum Unit p → TestResult p
+| gaveUp : Nat → TestResult p
+| failure : ¬p → List String → Nat → TestResult p
+deriving Inhabited
+
+/-- Configuration for testing a property. -/
+structure Configuration where
+  numInst : Nat := 100
+  maxSize : Nat := 100
+  numRetries : Nat := 10
+  traceDiscarded : Bool := false
+  traceSuccesses : Bool := false
+  traceShrink : Bool := false
+  traceShrinkCandidates : Bool := false
+  randomSeed : Option Nat := none
+  quiet : Bool := false
+  deriving Inhabited
+/--
+`PrintableProp p` allows one to print a proposition so that
+`SlimCheck` can indicate how values relate to each other.
+It's basically a poor man's delaborator.
+-/
+class PrintableProp (p : Prop) where
+  printProp : Option String
+
+export PrintableProp (printProp)
+
+instance (priority := low) : PrintableProp p where
+  printProp := none
+
+/-- `Testable p` uses random examples to try to disprove `p`. -/
+class Testable (p : Prop) where
+  run {} (cfg : Configuration) (minimize : Bool) : Gen (TestResult p)
+
+abbrev NamedBinder (n : String) (p : Prop) : Prop := p
```

Comment:
```suggestion
def NamedBinder (n : String) (p : Prop) : Prop := p
```
TC synthesis is supposed to see through reducible definitions, I'm not sure if we should count on this continuing to work.